reflex/utils/hosting.py

```diff
@@ -0,0 +1,1028 @@
+        from_iso_timestamp: An optional timestamp with timezone info to limit
+                            where the log queries should start from.
+
+    Raises:
+        ValueError: If the key is not provided.
+        Exception: If the operation fails. The exception message is the reason.
+
+    """
+    if not (token := authenticated_token()):
+        raise Exception("not authenticated")
+    if not key:
+        raise ValueError("Valid key is required for querying logs.")
+    try:
+        logs_endpoint = f"{constants.Hosting.DEPLOYMENT_LOGS_ENDPOINT}/{key}/logs?access_token={token}&log_type={log_type.value}"
+        console.debug(f"log server endpoint: {logs_endpoint}")
+        if from_iso_timestamp is not None:
+            logs_endpoint += (
+                f"&from_iso_timestamp={from_iso_timestamp.astimezone().isoformat()}"
+            )
+        _ws = websockets.connect(logs_endpoint)  # type: ignore
+        async with _ws as ws:
+            while True:
+                row_json = json.loads(await ws.recv())
+                console.debug(f"Server responded with logs: {row_json}")
+                if row_json and isinstance(row_json, dict):
+                    if "timestamp" in row_json:
+                        row_json["timestamp"] = convert_to_local_time(
+                            row_json["timestamp"]
+                        )
+                    print(" | ".join(row_json.values()))
+                else:
+                    console.debug("Server responded, no new logs, this is normal")
+    except Exception as ex:
+        console.debug(f"Unable to get more deployment logs due to {ex}.")
+        console.print("Log server disconnected ...")
+        console.print(
+            "Note that the server has limit to only stream logs for several minutes to conserve resources"
+        )
+
+
+def check_requirements_txt_exist():
+    """Check if requirements.txt exists in the current directory.
+
+    Raises:
+        Exception: If the requirements.txt does not exist.
+    """
+    if not os.path.exists(constants.RequirementsTxt.FILE):
+        raise Exception(
+            f"Unable to find {constants.RequirementsTxt.FILE} in the current directory."
+        )
+
+
+def authenticate_on_browser(
+    invitation_code: str | None,
+) -> tuple[str | None, str | None]:
+    """Open the browser to authenticate the user.
+
+    Args:
+        invitation_code: The invitation code if it exists.
+
+    Raises:
+        SystemExit: If the browser cannot be opened.
+
+    Returns:
+        The access token and invitation if valid, Nones otherwise.
+    """
+    console.print(f"Opening {constants.Hosting.CP_WEB_URL} ...")
+    request_id = uuid.uuid4().hex
+    if not webbrowser.open(
+        f"{constants.Hosting.CP_WEB_URL}?request-id={request_id}&code={invitation_code}"
+    ):
+        console.error(
+            f"Unable to open the browser to authenticate. Please contact support."
+        )
```

Comment:
hmm, that's a very good point. I do have a ticket to support multiple modes of login. We could support a pure command line auth. That's part of the plan in the future. We wanted to avoid getting into any password business, so oauth is definitely easier and less responsibility. Asking them to open the link probably won't work. Right now, the link has a query param `request-id`, this is how I used to identify who just tries to authorize then subsequently send them the token in the backend. This way they don't have to know about the token and just authenticate and stays logged in for a while.

Reply:
At this point in the code you already have the formatted URL because you just tried to send the browser there. It should have your request id param just in the URL, so if you print it, the user can copy/paste it with the id and the cli code can pull down the token in the same way.

Is there something I'm missing?

Reply:
@masenf I think I just overthought it. Yes, it'll be the same if we just ask the user to open that link manually (instead of existing) and still wait while they do their business.